docs/proposals/20231122-automate-aks-features.md

```diff
@@ -0,0 +1,342 @@
+
+It also has the drawback that one version of CAPZ is tied directly to a single AKS API version. The spec could
+potentially contain separate fields for each API version and enforce in webhooks that only one API version is
+being used at a time. Alternatively, users may set fields only present in a newer API version directly on the
+ManagedCluster after creation (if allowed by AKS) because CAPZ will not override user-provided fields for
+which it does not have its own opinion on ASO resources.
+
+While it couples CAPZ to one ASO API version, this approach allows CAPZ to move a more calculated pace with
+regards to AKS API versions the way it's done today. This also narrows CAPZ's scope of responsibility which
+reduces CAPZ's exposure to potential incompatibilities with certain ASO API versions.
+
+Regarding ClusterClass, this option functions the same as [Option 2] or [Option 3], where all ASO fields can
+be defined in a template. This option opens up an additional safeguard though, where webhooks could flag
+fields which should not be defined in a template. Similar webhook checks would be less practical in the other
+options where CAPZ would need to be aware of the set of disallowed fields for each ASO API version that a user
+could use.
+
+#### Decision
+
+We are opting to move forward with [Option 4]. That approach is most consistent with how the API is defined
+today and would make for the smoothest transition for users.
+
+### Security Model
+
+Overall, none of the approaches outlined above change what data is ultimately represented in the API, only the
+higher-level shape of the API. That means there is no further transport or handling of secrets or other
+sensitive information beyond what CAPZ already does.
+
+### Risks and Mitigations
+
+Increasing CAPZ's reliance on ASO and exposing ASO to users at the API level further increases the risk
+[previously discussed](https://github.com/kubernetes-sigs/cluster-api-provider-azure/blob/ce3c130266b23a8b67aa5ef9a21f257ff9e6d63e/docs/proposals/20230123-azure-service-operator.md?plain=1#L169)
+that since ASO has not yet been proven to be as much of a staple as the other projects that manage
+infrastructure on Azure, ASO's lifespan may be more limited than others. If ASO were to sunset while CAPZ
+still relies on it, CAPZ would have to rework its APIs. This risk is mitigated by the fact that no
+announcements have yet been made regarding ASO's end-of-life and the project continues to be very active.
+
+## Alternatives
+
+The main alternative to leveraging ASO to expose the entire AKS API from CAPZ would be for CAPZ to build its
+own code generation pipeline which takes as input the Azure API specs and ultimately produces definitions for
+AzureManagedControlPlane and AzureManagedMachinePool and mappings to ASO's ManagedCluster and
+ManagedClustersAgentPool. Such a solution would likely end up being too similar to what ASO already provides
```

Comment:
Agree w/ this - I would avoid code generation and attempting to read the Swagger specifications (or SDKs) unless you really have to. It really is nontrivial to get it right (ASO struggles even now for some service Swaggers that use funky extensions or have weird type structures).

If you really end up not needing/wanting ASO for whatever reason, I would still probably just use the ASO code generator directly and we can give you a mode to it to produce the types you need. It has the infrastructure to do this and in a perfect world what it is is an "Azure Swagger -> CRD" code generator and not necessarily something that _has_ to produce ASO specific shapes (though in practice now that's mostly what it does produce).